File: LeetCode/cpp/131.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        int n = s.size();
        if (n == 0)
            return ret;
        vector<string> temp;
        BackTracking(s,n,0,temp);
        return ret;
    }

    void BackTracking(string s,int &n,int i,vector<string> &temp){
        if(i == n) {
            ret.push_back(temp);
            return;
        }
        for(int j =i ; j < n; ++j) {
            if(isPalindrome(i,j,s)) {
                temp.push_back(s.substr(i,j-i + 1));
                BackTracking(s,n,j+1,temp);
                temp.pop_back();
            }
        }
    }

    bool isPalindrome(int i, int j,string &s) {
        while(i <= j &&s[i] == s[j]) {
            ++i;
            --j;
        }
        if(i > j)
            return true;
        return false;
    }

private:
    vector<vector<string>> ret;
};

File: LeetCode/python3/131.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        ret = []
        self.Backtrack(s, 0, [], ret)
        return ret

    def Backtrack(self, s, k, curr, ret):
        if k == len(s):
            ret.append([c for c in curr])
            return curr
        for i in range(k, len(s)):
            if self.IsPalindrome(s, k, i):
                curr.append(s[k:i + 1])
                curr = self.Backtrack(s, i + 1, curr, ret)
                curr.pop()
        return curr

    def IsPalindrome(self, s, i, j):
        while j > i:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1
        return True
